### ISSUE TITLE
OPA's WASM runtime leaks memory when incrementally adding or removing data

### ISSUE THREAD

**Issue by ctelfer-sophos** [open-policy-agent/opa#5785]
## Short description 
Greetings,

This is a continuation of the community discussion here: https://github.com/orgs/open-policy-agent/discussions/359

With the current version of OPA (tip as of last week: eb3cf9096) as well as with previous versions of OPA (at least to commit 539543ed01 from October 2020), it seems that the OPA WASM VM will leak memory over time when when the host program invokes `opa_value_add_path()` and/or `opa_value_remove_path()` in combination with `opa_eval()`.  From what I can see, this is caused by a collection of issues relating to a combination of shallow frees and re-initialization of the heap.

I came across this while looking at embedding the OPA WASM engine in another application.  However, the issue will also manifest in the OPA agent running in server mode if that server is using WASM VMs internally and when data documents are being incrementally updated.  However, I haven't attempted to reproduce the server scenario yet. 

## Steps To Reproduce
These steps are from the perspective of running OPA in a WASM engine in a local program.   Just about any policy will work.  Something as simple as:

```
package result
foo := { "bar" : "xyzzy" }
allow { foo.bar }
```

1. Build the policy to wasm by running `opa build -t wasm policy.rego`
2. Extract the WASM (e.g. `DIR=$(pwd) ; mkdir -p /tmp/foo; mv bundle.tar.gz ~/tmp/foo ; pushd /tmp/foo ; tar zxvf bundle.tar.gz ; cp policy.wasm $(pwd) ; cd .. ; rm -r foo ; popd`  )
3. Initialize an OPA VM as per https://www.openpolicyagent.org/docs/latest/wasm/#from-scratch and load it with the policy.wasm
4. Get the initial heap pointer by invoking `opa_heap_ptr_get()`
5. Repeatedly do the following
    * Following the conventions here (https://www.openpolicyagent.org/docs/latest/wasm/#external-data) also found in the OPA code here (https://github.com/open-policy-agent/opa/blob/main/internal/wasm/sdk/internal/wasm/vm.go#L522) add the value "aaaaaaaaa" to the path "data.result.foo.bar".  You can increment the value string or leave it the same.  It doesn't matter.
    * Following (https://www.openpolicyagent.org/docs/latest/wasm/#evaluation) invoke `eval()` to query the data.  Or use `opa_eval()` as in https://www.openpolicyagent.org/docs/latest/wasm/#exports.  I prefer this
    * Get the new heap pointer via `opa_heap_ptr_get()` and compare with previous versions

You will see the heap pointer steadily rising.  It won't matter if you insert the same string over and over.  It also won't matter if you invoke `opa_value_remove_path()` with every invocation.

There is probably a similar way to reproduce this using the agent.  There you'd just have to watch the RSS of the process with top over time and lots of invocations.  But I'm unclear when exactly WASM VMs are present and/or updated when running in server mode.

## Expected behavior
Given enough invocations, the heap pointer should eventually stablize as we shouln't be changing the amount of memory needed for evaluation.  An initial spike makes some sense.  But this will leak data continuously.

## Additional context
I've done a lot more diving into this since the initial examiniation.  I'll post a full write-up of my triage with more details after filing the initial the issue.  I don't want the summary to be too cluttered up.  

In the meantime, however, I'm copying the following from the community discussion for completeness.

### Community Discussion Info Shared

The documentation (https://www.openpolicyagent.org/docs/latest/wasm/#exports) seems to indicate that "Values removed will be freed". However, running repeated remove()/add()s in a tight loop ends up leaking memory very quickly. I've readily observed this in test code (repeated add()/remove() using same path and same data).

From what I can tell from looking at the source (https://github.com/open-policy-agent/opa/blob/main/wasm/src/value.c), the issue seems to come down to shallow `opa_free()`s in several places.

* The documentation indicates to use an `opa_free()` on the parsed path argument for `opa_value_add_path()`, but this will only free the array object at best. It won't free the elements, their index keys or the strings they'd point to.
* The remove operation (as well as the add on a path holding other objects) invokes the internal function `opa_value_free()` which will do some freeing of the pointed to objects, but not a deep free. For example, if the object pointed to were another object, `opa_value_free()` would release both the object and its buckets. But it wouldn't attempt to free anything the object referred to.

Although it would be painful, one might be able to prune the data bit by bit using `opa_value_remove_path()` starting from the leaves of the data and working back. But that wouldn't address the issue of the "path object" also leaking memory.

### Summary of Additional Findings After Discussion

There are several other memory leaks as it turns out in the system.  The most significant is that every time one invokes `opa_eval()`, the WASM engine invokes `opa_heap_set_ptr()` which resets the heap and leaks all free blocks in the freelist.  Since invoking `opa_value_add_path()`/`opa_value_remove_path()` involves allocating other buffers that will get released to the freelist (but without decreasing the heap pointer because value nodes added to the tree will be at higher addresses in the heap).  Subsequent `opa_eval()` calls will then leak these.  

Another issue is that when adding data to the data document, the `opa_object_insert()` call can leak keys.

**Comment by ctelfer-sophos** [comment]
As promised, here's my more detailed analysis.

## OPA WASM VM Memory Layout
From comments in the WASM, the memory layout with the WASM VM is:

     //   |  <-- stack | -- data -- | heap -->  |

The “data” portion in this refers to the policy data.  More data can be added by allocating it from the heap and linking it into the existing data tree.
OPA manages its heap through the functions `opa_malloc()`, `opa_free()` and `opa_heap_ptr_set()`.  These manage the heap data using a simple, direct, bucket-based allocator.  The allocator keeps 5 lists of free “blocks” of differing sizes.  The first four lists contain 4-, 8-, 16- and 64-byte blocks each prefixed by a common 12-byte header for management. The fifth list contains variable-sized blocks and each such block is allocated to a minimum of 128 bytes. The `opa_malloc()` function will attempt to satisfy each memory request from a block on the list with the smallest size necessary to hold the requested data.  Each block list contains the list of available blocks stored in address order.   The allocator does not attempt to perform any alignment of the heap data it allocates.  A single global “heap pointer” points to the next address to allocate heap data from that is not stored in any block list.  
The there is also a “top pointer” that points to the very top of the heap. If the `opa_malloc()` function needs to allocate memory, but the heap pointer is at or would have to pass the “top pointer”, then it invokes an intrinsic WASM function `__builtin_wasm_memory_grow()` to increase the top of the heap.  The WASM runtime fulfills or denies this request.  If this function fails, `opa_malloc()` invokes the import function `opa_abort()` to abort the WASM program being run.  There is no recovery possible from `opa_abort()`.  So if an OPA WASM runs out of memory performing some operation, the WASM VM is left in an unrecoverable state. 

When `opa_free()` releases data, it looks at the (prepended) 12-byte header to find the size of the block being freed and from this decides which block list it belongs to.  It then re-inserts the block into its appropriate block list.  If the block list is the variable sized (i.e. the last block list) then it will attempt to coalesce adjacent blocks in memory.  However, it never attempts to coalesce fixed-size blocks.  After returning the block to the list (and coalescing if applicable), `opa_free()` will check whether the block ends at the heap pointer.  If so, it will remove blocks from the block list and decrease the heap pointer.  It will do this repeatedly as there may be many adjacent free blocks near the end of the heap.  The `opa_free()` function will stop releasing blocks and decreasing the heap pointer when the block list is empty or when the last block on the list doesn’t end at the heap pointer.   This may happen because the heap data ending at the heap pointer is in an allocated block, or it may happen because that block is on another block list.  The function does not attempt to traverse other block lists to look for more data to release.

There is a second way to release data in the OPA WASM and that is to invoke the `opa_heap_ptr_set()` call.  This call tells the OPA allocator to reset the heap pointer to a fixed address (below the heap top).  The current implementation does this by resetting all the block lists to empty lists and setting the shared heap pointer to the aforementioned value.  This basically regards all data less than the new heap pointer as allocated.  If there were any blocks on the freelist below the new heap pointer, they are effectively leaked.

## `opa_eval()`
The `opa_eval()` call is a WASM export function and is the main entry point for OPA policy queries.  (There is also an older `eval()` call that one can use that requires calling multiple other OPA exported functions making overall query time slower.) One of the things that seems apparent in the design of this function is that it gets some speed from “arena-based” allocation.  That is, it allocates data structures incrementally from a common “arena” but can free it all with a single operation.  In this case, the “arena” is all heap memory above the heap pointer as it existed just before the evaluation.

To invoke `opa_eval()` the caller must:
1. Ensure that there is room for the policy input (in OPA-JSON form) on the heap starting at the heap pointer.
2. Copy the input to the shared memory space.
3. Invoke `opa_eval()` passing the address of the policy data, the query entrypoint (i.e. which rule to query), the address of the input, the input length, the new heap pointer (which is usually the input + input length), and a boolean as whether to output the result in JSON or internal rego format.

In turn, the opa_eval() call then:
1. invokes `opa_heap_ptr_set()` on the new heap pointer.
    a. Note that this effectively frees all heap data at an address preceding the input JSON
2. parses the input JSON into internal data structures
3. sets up internal data structures and invokes the OPA query (which the OPA compiler builds from rego)
4. generates JSON (or internally-formatted) output
5. returns the address of this output to the caller
    a. This is a null-terminated byte string, so the length is implicit

Note that the caller will usually invoke `opa_heap_ptr_get()` after loading policy data into the VM and use that as the starting address for input to all calls to `opa_eval()`.  This is safe because `opa_eval()` will never attempt to allocate data below the boundary of input + input_length due to its call to `opa_heap_ptr_set()`.  Since `opa_eval()` invokes `opa_heap_ptr_set()` with every call, it frees any allocated data from the previous `opa_eval()` call.  The `opa_eval()` call never frees any of the policy data.

## Setting Policy Data

When OPA compiles a WASM module, it embeds the policy data that it knows about into the WASM logic and into internal data structures.  (This is speculation, but I think my guess is a good one.) When one invokes `opa_eval()`, one of the parameters to that call is the offset of the “data document".  If the caller does not provide new policy data as part of the call, then they pass a NULL (0) for the data document and the internal `eval()` call uses the data tree that the compiler built during compilation.  (Or is it just that the initial data document happens to be at address 0?). However, the caller has the option to set their own data tree instead of any that the compiler built into the WASM.

To do this the caller:
1. Set the heap pointer to a known “safe” address (call it the base heap pointer).
  a. Usually during VM initialization, the caller invokes opa_heap_ptr_get() after an opa_malloc(0) call to get the pointer to the initial heap pointer.   A call to opa_heap_ptr_set() with this value will effectively reset the heap to the state it was in when the VM first initialized (i.e. right after the initial WASM was loaded).
2. Invokes opa_malloc() to allocate a buffer to hold the JSON form of the data.
3. Copies the JSON data into the buffer (using APIs from the WASM runtime).
4. Invokes the export `opa_value_parse()` to convert this value into OPA’s internal data representation
    a. This function returns the address of the new policy data
5. Stores the return address from `opa_value_parse()` to pass to the next invocation of `opa_eval()`
6. Invokes `opa_free()` on the JSON buffer to release the JSON data.
7. Invokes `opa_heap_ptr_get()` to get a new “heap pointer”.  This is the location where the input will go on the next call to `opa_eval()` (see above).

If the caller ever wishes to discard this data they can invoke `opa_heap_ptr_set()` on the original base heap pointer found during VM initialization and return to passing 0 for the data offset.

It should be noted that although the caller invokes `opa_free()` to release the buffer for the JSON data, this is unlikely to save any memory and will likely leak the buffer.  This is because the allocator will most likely have no free blocks before setting the policy data.  Therefore the sequence above will allocate space for the JSON data first and the blocks for the parsed data tree second.  Therefore, the parsed data tree will be at higher addresses than the JSON buffer.  Based on the internals of `opa_free()` above, `opa_free()` will return the JSON buffer to the heap, but the heap pointer will not decrease since there are still allocated blocks at higher addresses.  When the `opa_eval()` call eventually invokes `opa_heap_ptr_set()` (see above) this will cause the JSON buffer which is at a lower address to leak.

Note that while this data leak is undesirable, it is not catastrophic.  Continued use of `opa_eval()` will not leak any further data.  Furthermore, resetting the policy data entirely by invoke `opa_heap_ptr_set()` to the base heap pointer will reclaim both he parsed policy buffer and the JSON buffer.

## Incremental Policy Data Addition and Removal
OPA’s WASM VM provides two functions for manipulating policy data dynamically.  The `opa_value_add_path()` export function can add policy data.  The `opa_value_remove_path()` export function can remove policy data.  Both functions take a “path” argument.  In the case of `opa_value_add_path()` this is the path through the data tree to the location to place the new data.  In the case of `opa_value_remove_path()` this is the path through the data tree to the subtree to remove.  The `opa_value_add_path()` will add intermediate nodes in the data tree if they aren’t present when traversing it to the insertion point.

In order to call `opa_value_add_path()` the caller:
1. Invokes `opa_heap_ptr_set()` to clear any lingering data from previous `opa_eval()` invocations
2. Calls `opa_malloc()` to allocate a buffer for the JSON form of the data to add
3. Copies the new data into the buffer (via the WASM runtime)
4. Calls `opa_value_parse()` to parse the data into OPA’s internal representation
    a. This function returns the address of the root of the subtree of data to add
5. Calls `opa_free()` to release JSON buffer
6. Calls `opa_malloc()` to allocate space for the path in JSON form
    a. The path must take the form of JSON array of strings.
    b. Each string in the JSON array is another object field or array index to traverse to get to the insertion point.
7. Copies the path into the JSON buffer
8. Call `opa_value_parse()` to convert the path to the internal representation
9. Calls `opa_free()` to free the JSON buffer of the path
10. Calls `opa_value_add_path()` 
11. Calls `opa_free()` on the parsed path to “free” it. (but see below)
12. Calls `opa_heap_ptr_get()` to get a new safe base pointer for calls to opa_eval().
    a. Again, this is where the caller to `opa_eval()` will start copying input.

The `opa_value_add_path()` call:
1. Walks the data tree to find the subtree specified by path
    a. As mentioned above, it will insert “objects” (dictionaries) in the tree for missing elements in the path
2. Saves the old subtree at the point of the path
3. Inserts the new data as a subtree to the last node in the path
4. “frees” the old subtree (again, see below)

The `opa_value_remove_path()` calling convention is very similar.  The differences are:
1. It omits the step of allocating and parsing JSON insertion data (there is none) and freeing the JSON buffer
2. The caller doesn’t typically invoke opa_heap_ptr_get() afterwards.

The function itself traverses the data tree, but will not add any missing nodes.  Indeed, if a path element is missing in the tree, the function will merely stop and return success.  Obviously, the remove function will also not insert policy data.  It just removes the subtree.

## Leaks in `opa_value_[add|remove]_path()`
From inspection of the WASM code, there are two immediate and direct data leaks.  First, when freeing the internal representation of a data path, the documentation says to invoke `opa_free()`.  However, the address of the path value is a pointer to an `opa_array_t` object which in turn points to an array of `opa_array_elem_t` objects.  Each `opa_array_elem_t` object points to both an index object (in this case an `opa_number_t`) and a value object (in this case, an `opa_string_t`).  Invoking `opa_free()` merely frees the `opa_array_t` structure and none of the other memory that it refers to.  Their memory is, therefore, leaked.

In the second place, when either `opa_value_add_path()` or `opa_value_remove_path()` tries to free the old value stored at the path, they invoke the internal function `opa_value_free()`.  This function looks at the type of the value being freed and does a little better.  If that value is a boolean, number or string, the free will be successful.  Unfortunately, if the value is another object, or array or set, `opa_value_free()` will release (using `opa_free()`) some of the data, but will not attempt to traverse deeper into the subtree to release other referenced values.  For example, if an object contains another value, `opa_value_free()` will release the outer `opa_object_t` along with the array of `opa_object_elem_t` bucket pointers and the `opa_object_elem_t` buckets themselves for that outer object.  But it will leak the inner value along with any values that it refers to. The same behavior applies to arrays and sets.

Now, it might be possible to avoid the second leak by invoking `opa_value_remove_path()` starting with the leaves of the data subtree and working backwards.  This would be expensive and complicated, but it would be possible.  Unfortunately, that option is made moot by the first leak.
  
One might cleverly think, that one could invoke `opa_value_remove_path()` on the elements of the path object as well to correctly free it, but this is not possible because the path must always contain at least one element and so one can’t free the base array object this way.  It’s also a circular dependency since one must have a path to remove a path.

## Fixing the Memory leaks

### Addressing the Lack of “Deep” Free

#### Freeing the Path
Safely supporting the incremental add/remove of policy data requires, first and foremost, the ability to correctly release the path array used in the add/remove builtins.  Some potential approaches to this include:
* having the add/remove_path functions themselves parse the path string into a value and properly release the data
* having the add/remove_path function parse the string, but NOT into a value enabling a direct free to function properly
* exposing a new builtin that the host software can invoke to correctly free the path value
The last could arguably be the best because it also helps address the second issue.  

#### Cleaning up the Removed Subtree
Given a safe way to free the path, the host program could perform safe incremental data addition/removal by doing depth-first pruning of those trees.  But this would be very expensive in terms of the number of WASM invocations required.  It would also require the host program to keep a current copy of the data by which to reconcile the removals.  This isn’t an exceptional burden on the OPA agent which already does this.  Finally, this approach requires the host program to perform the depth-first search of the subtree.  It’s generally poor form in an API to require such a complex operation to use an API according to basic safety.

The most programmer-friendly way to address the second concern, therefore, is to enable OPA’s WASM engine to “deep free” the subtree of policy objects in the user data.  This is readily feasible if the structures in the data document form a true tree (i.e. a graph with no cycles). The `opa_value_free()` function invokes type-specific free functions to ensure the release of auxiliary data based on the type of value.  We could change the `opa_array_free()`, `opa_object_free()` and `opa_set_free()` calls to release other values that they refer to by invoking the `opa_value_free()` call (effectively making the operation recursive).  

There is good news and bad news here.   From my examination of the code, it looks like the data document itself forms a tree.  However, the `eval()` function that implements queries layers other values on top of this tree creating a directed acyclic graph (DAG) effectively.  The problem with a DAG would be that two values could never to a common value.  (e.g. two set values could have “value” pointers that point to the same value data structure). A recursive free of one of those sets would leave the other with a dangling pointer.

However, if that original observation holds true (that the data document itself is a tree), then this offers a bit of hope because OPA doesn’t release `eval()` query memory using `opa_value_free()` (mostly).  Generally, it frees memory by resetting the heap pointer between queries implicitly freeing all the allocated memory above the heap pointer.

So far, I have found two exceptions to the above rule.  The function `opa_value_compare_object()` creates temporary arrays with references to keys but then invokes `opa_array_free()` to release the arrays.  If we modify `opa_array_free()` to naively do a deep free as per above, it will end up freeing the keys as well.  Similarly, `opa_set_intersection()` and `opa_set_union()` invoke `opa_value_free()` to free shallow copies of sets referring to common underlying value members.  So making `opa_value_free()` naively deep copy will, again, produce dangling pointers for these functions as well.

It is possible to address both of these issues simultaneously, by making the application of a deep free on objects, sets and arrays optional by means of a new parameter (flag).  For example, we can make `opa_array_free()` take a second parameter indicating whether to recursively invoke `opa_value_free()` on its elements.  Similarly, we can make `opa_value_free()` take a second parameter which it passes to its value-type-specific sub-functions about whether to perform a deep free or not.  In the interest code hygiene, we’d probably rename the function to be `__opa_value_free()` taking two parameters and create two wrapper functions that invoke the shallow or deep copy specifically.  The `opa_set_intersection()` and `opa_set_union()` could invoke the shallow version since those functions don’t manipulate the data document, but rather only intermediate results used during queries.  Other functions would invoke the deep version.

Finally, if we now have an appropriate version of `opa_value_free()` that performs a deep free (of the data document), then we could expose this as a new WASM export that the host program can invoke on the path value above in order to correctly free its memory.

#### Another Leak: `opa_object_insert()`
Even with the above changes, there is one further memory-related issue that can arise.  The function `opa_object_insert_path()` invokes `opa_object_insert()` to add a new subtree into the data document.  Unfortunately, `opa_object_insert()` is inconsistent in how it consumes the key structure that it receives as a parameter.  If the object already has a value at the same key, `opa_object_insert()` simply sets the value pointer in the bucket node to the new value.  It does not free the key passed in as a parameter.  (note, this is not a leak for `opa_object_insert_path()` as it will invoke `opa_value_free()` on the old value). However, if the key does not exist in the object, then `opa_object_insert()` allocates a new bucket and sets it’s “key pointer” to point to the new key and the “value pointer” to the new value.  The caller gets no indication of whether the key they passed in was added to the object tree or not.  They just assume that it was.

We could address this by somehow passing the “you need to free the key” information back to the caller or by freeing one of the keys (the original or the new one) if there is a duplicate.  Unfortunately, `opa_object_insert()` is used in many places where the key being passed in is not owned solely by the caller.  (e.g. the key is also referred to by some other object). So this approach probably requires a split API similar to the one mentioned for `opa_value_free()` with a parameter coming to the actual implementation indicating whether it should free the duplicate key or not.

One further issue is that `opa_object_insert_path()` passes a shallow reference to the path element to `opa_object_insert()`.  So, if the user eventually frees the path using a deep free, they’ll leave a dangling reference in the data document.  The `opa_object_insert_path()` could address this either by pruning the path element from the array or by duplicating the key (which is just an `opa_string_t`) before calling `opa_object_insert()`.

### Addressing `opa_heap_ptr_set()` Leaks

#### A Wrong Approach To Learn From
All of the aforementioned fixes will not prevent `opa_eval()` from leaking any free space on the heap when it invokes `opa_heap_ptr_set()`.  I had originally tried modifying `opa_heap_ptr_set()` so that instead of releasing all objects in the freelist and setting the heap pointer, it would walk the freelists starting from the latest address until reaching `heap_ptr` and removing heap blocks from those lists.  The allocator would allocate said memory from the heap pointer anyway.  This wouldn’t be as fast as the arena-based memory freeing, but it would theoretically allow the WASM VM to keep reusing the memory allocated below the “safe” heap pointer.

Unfortunately, while this approach is a significant improvement, it still ends up leaking memory over time.  This is because when `opa_eval()` starts using the heap memory, it will allocate blocks from below the heap pointer (since they’re available).  When `opa_eval()` on a subsequent call invokes `opa_heap_ptr_set()`, the memory allocated above the base heap pointer gets reclaimed.  But the memory below that heap pointer, however is still allocated and thus it gets leaked.

If one were just querying from then on for the rest of the program, eventually all the “below heap base pointer” memory would get leaked and then new `opa_eval()` invocations would only allocate memory from above the base heap pointer and that memory would always be reclaimed successfully.  So the leak would be limited.   But that would only hold true up until the host program attempted to call `opa_value_[add|remove]_path() again`.  This will leave more memory below the base heap pointer.  Most likely, the base heap pointer will shift up to accommodate new data while buffers for path json, path value and data json are all below the heap pointer.  Eventually `opa_eval()` will leak these too.

The net/net is that a series of `opa_eval()` and `opa_value_[add|remove]_path()` invocations will still leak memory over time although at a much slower rate.

### Giving Up on Letting `opa_eval()` Use Heap Memory Below the Base Heap Pointer
However, the approach above did lead me to an important realization.  In its current form, `opa_eval()` is going to only use memory above its base heap pointer.  (It leaks all space below it.)  Even there are free blocks on the freelist, `opa_eval()` would leak them eventually.  So there’s no point in saving said memory for `opa_eval()` to use.  But we could save that memory for further data document manipulation operations.

With this approach, before we invoke `opa_eval()` for the first time we stash all the heap memory on the freelists in a shadow set of freelists.  When `opa_eval()` resets the heap pointer, it won’t leak any data since the freelists will be empty.  This can be done in O(1) time per freelist by moving the freelist in bulk to the shadow freelists.  When `opa_eval()` runs it will only use memory above the base heap pointer which is the same behavior as today.

However, when a program wants to change the data document, it first restores the stashed heap blocks to the freelists.  Next it makes any changes to the data document allowing its allocations to use the stashed heap memory.  Finally, once the modifications are complete, the program again stashes any heap blocks on the freelists back to the shadow lists.  There may be more or less blocks depending on what the host program added, replaced and/or removed from the tree.  But the overall storage won’t be wasted.

This approach is completely backward compatible with he current solution since existing behavior won’t change.

## Summary
As of yet, I don't see a good approach to avoiding these memory leaks without changing the WASM API.  Properly freeing OPA "values" requires access to `opa_value_free()`.   Furthermore, `opa_eval()` would require modification to free transient data and free it intelligently.  Not only would this be a very fundamental change to the object model, it would likely drastically impact performance.

I have a backwards-compatible API change that I'll be happy to share via PR that seems to address the issue.  It follows the approach described above.  In my tests so far, I can run 100,000 iterations of `opa_value_set_data()` followed by `opa_eval()` without growing the stack by a single byte except for during the first iteration.

But first, it would probably be better to hear from the rest of the community whether these analyses are correct and/or whether there are other potential solutions.

**Comment by ctelfer-sophos** [comment]
Updated the script to parameterize a few things [opa-leak.tgz](https://github.com/open-policy-agent/opa/files/11125655/opa-leak.tgz)

You can now run as: 
    `$ OPA=/path/to/opa NREPS=<# of reps per test> TARGET=[wasm|rego|plan] ./test-leak.sh`
So you can set the path to your binary and the # of reps per test.  You can also set whether to build the policy for WASM, Rego or a "plan" (serialized IR?).

Using "rego" as a target I confirmed that with the same binary I'm not seeing memory increase with the patch/query/patch/query cycles.
```
...
Running patch/query/patch/query in loop
500 sets of patch/query/patch/query requests
mem info: 337745 8721 6279 5515 0 30087 0
...
10000 sets of patch/query/patch/query requests
mem info: 337745 8757 6324 5515 0 30087 0

Stopping server
```

**Comment by ashutosh-narkar** [comment]
@ctelfer-sophos can you please take a look into [this](https://github.com/open-policy-agent/opa/actions/runs/4834590354/jobs/8616094204#step:3:790) test failure. One of the benchmark tests seems to be time out post merging your PR. Do you have an idea what would have caused this?

**Comment by ctelfer-sophos** [comment]
Ouch.  Ok, full disclosure, I have had this exact test lock up on me with a timeout with _unmodified_ OPA before.  I've been meaning to ask about that.  BUT, having said that, from the stack trace it looks like most of the goroutines are idle *except* for one that's invoking the SetDataPath() / opa_value_add_path() function which is freezing in a "system call".  So, that would point to a bug in my changes causing some kind of infinite loop or some such.

I'll try to reproduce and further debug.  Is there an easy way to run only this test?  Looks like it would be something like: 
```
CGO_ENABLED=1 GOFLAGS="-buildmode=exe" go test -tags=opa_wasm,slow -timeout 30m -run=- -bench=SOME_PATH??? -benchmem ./SOME_PATH???
```
But trying the paths that make sense doesn't seem to run the benchmark in question.

**Comment by ctelfer-sophos** [comment]
Sorry for the dumb question. I forgot that the --bench= argument takes a regex.  So far my testing hasn't indicated a loop.  This test takes a long time to run in general and it might be that the changes push that runtime over the 25 minute limit.  If so, that would mean this change causes a perf regression which may not be that surprising given that it does extra work per update.  Or it might be something worse as previously feared.  I'll keep experimenting and testing and report what I find...

**Comment by ctelfer-sophos** [comment]
Update: 
* running the benchmark individually completes without issue
* running the e2e wasm benchmarks complete without issue
* running the e2e benchmarks (wasm & non-wasm) complete without issue
* running the full 'make perf' I've hit the freeze locally
* running the test/e2e/authz tests I've seen the ns/op being 50% higher with the fix vs without
* **BUT** that may be my test environment because subsequent runs have produced numbers consistent with the pre-patch numbers.  

So, so far I've only been able to reproduce w/ a full 'make test' (and that takes quite a bit of time before it hits the test in question :( )

**Comment by ashutosh-narkar** [comment]
Thanks for helping to debug the issue.

**Comment by ctelfer-sophos** [comment]
Further note.  Have run with a debug build detecting double-frees.  Hit the timeout, but there were no double-frees detected.  So that eliminates any loops caused by double-free as a cause.  Will start looking closer at instrumenting the deep free code to see if there is a recursive loop happening there.  (but think that would stack overflow if it did, so that doesn't seem likely...)

**Comment by ctelfer-sophos** [comment]
Ok, more progress(-ish) to report.  Have tried a build that detects any circularity in the opa_value_free() recursion.  (WASM engine set to abort if detected.)   As above, all builds succeed and all tests pass and was able to trigger the timeout, but not the circularity was detected.  This build also detected overlong string copies and path lengths.  (Basically, looking at the loops that opa_value_path_add() invokes to see if any are going haywire.). None of these assertions (opa_abort()s) triggered.

So, still not seeing anything pointing to an infinite (or super-long running) loop of any sort...

**Comment by ashutosh-narkar** [comment]
@ctelfer-sophos  any more updates on finding the issue?

**Comment by ctelfer-sophos** [comment]
I don't think I'm much closer to a root cause yet.  If you'd prefer to revert the change so others can go in, that would probably make sense at this point.

A few more observations:
* Normally when running the performance tests (including the authz tests when running in e2e mode) I see two different processes taking up ~99% CPU.  But when I observe the lockup I see only one.
* This leads me to  think that the other process either stalled completely (no CPU, waiting in the kernel somehow) or never started in the first place.
* So I'm going to look at ways to instrument both the WASM and the go test code to count how many test iterations and/or opa_data_add_path() calls take place during that particular test.
* Hopefully, from this, I'll be able to determine whether this is a complete lock-up (as it appears) or just progressing super-slowly.
* Still haven't found any way to reproduce short of running the full perf suite.   Other combinations of perf benchmarks simply won't trigger the stall observed during the complete benchmark so far.

I'll also post the patches I used to gather the previous results for reference.

**Comment by ashutosh-narkar** [comment]
Thanks for looking into this @ctelfer-sophos. Would you be able to create a new issue with your findings so that we can track this. Also link this issue in there for reference. For now I'm planning to skip that particular benchmark to unblock the workflow.

**Comment by ctelfer-sophos** [comment]
@ashutosh-narkar Sure.  Will do this later this evening.
